contractcourt/chain_watcher.go

```diff
@@ -974,24 +977,62 @@ func (c *chainWatcher) handleUnknownRemoteState(
 // return zero. This is possible as our output may have been trimmed due to
 // being dust.
 func (c *chainWatcher) toSelfAmount(tx *wire.MsgTx) btcutil.Amount {
-	var selfAmt btcutil.Amount
-	for _, txOut := range tx.TxOut {
+	// There are two main cases we have to handle here. First, in the coop
+	// close case we will always have saved the delivery address we used
+	// whether it was from the upfront shutdown, from the delivery address
+	// requested at close time, or even an automatically generated one. All
+	// coop-close cases can be identified in the following manner:
+	shutdown, _ := c.cfg.chanState.ShutdownInfo()
+	oDeliveryAddr := fn.MapOption(
+		func(i channeldb.ShutdownInfo) lnwire.DeliveryAddress {
+			return i.DeliveryScript.Val
+		})(shutdown)
+
+	// Here we define a function that can pick out the TxOut that
+	// corresponds with our local delivery script from a ShutdownInfo if we
+	// have a ShutdownInfo for this chainWather's channel.
+	//
+	// isDeliveryOutput :: *TxOut -> bool
+	isDeliveryOutput := func(o *wire.TxOut) bool {
+		return fn.ElimOption(
+			oDeliveryAddr,
+			// If we don't have a delivery addr, then the output
+			// can't match it.
+			func() bool { return false },
+			// Otherwise if the PkScript of the TxOut matches our
+			// delivery script then this is a delivery output.
+			func(a lnwire.DeliveryAddress) bool {
+				return slices.Equal(a, o.PkScript)
+			},
+		)
+	}
+
+	// Here we define a function capable of identifying whether an output
+	// belongs to the LND wallet. We use this as a heuristic in the case
```

Comment:
I am not sure I understand this comment, seems to me we currently only use this function in the cooperate close case, why do we mention the force closure output here ?

Reply:
While I hadn't looked at this prior to you pointing this out, it *does* appear that this function's only use site is in a context where we already assume/know it to be a coop close. That said, given that the function is named `toSelfAmount`, I think it's fair to keep it general enough to handle either case.

Generally, since the logical dependency relationship is that the caller depends on the callee, I don't think it's reasonable to assume how a function might be called. Perhaps if we want to limit it I can rename it to `coopToSelfAmount` and then trim out any commentary about how it might be used in a FC case. Perhaps this is the right compromise.